pandas/core/indexes/base.py

```diff
@@ -5233,16 +5237,32 @@ def get_indexer_for(self, target, **kwargs) -> np.ndarray:
 
         Returns
         -------
-        numpy.ndarray
+        numpy.ndarray[np.intp]
             List of indices.
         """
         if self._index_as_unique:
             return self.get_indexer(target, **kwargs)
         indexer, _ = self.get_indexer_non_unique(target)
         return indexer
 
+    @overload
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: Literal[True] = ...
+    ) -> np.ndarray:
+        # returned ndarray is np.intp
+        ...
+
+    @overload
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: Literal[False]
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
+        ...
+
     @final
-    def _get_indexer_non_comparable(self, target: Index, method, unique: bool = True):
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: bool = True
+    ) -> np.ndarray | tuple[np.ndarray, np.ndarray]:
```

Comment:
This return type needs to be in an overload, not in the function signature itself - see this example:
```python
from typing import Literal, overload, Optional

@overload
def func(inplace: Literal[True]) -> None: ...

@overload
def func(inplace: Literal[False] = ...) -> int: ...

def func(inplace: bool = False) -> Optional[int]: ...

inplace: bool
reveal_type(func(inplace=inplace))
```
gives us
```python
main.py:12: error: No overload variant of "func" matches argument type "bool"
main.py:12: note: Possible overload variants:
main.py:12: note:     def func(inplace: Literal[True]) -> None
main.py:12: note:     def func(inplace: Literal[False] = ...) -> int
main.py:12: note: Revealed type is 'Any'
Found 1 error in 1 file (checked 1 source file)
```

However, if we make a separate overload:
```python
from typing import Literal, overload, Optional

@overload
def func(inplace: Literal[True]) -> None: ...

@overload
def func(inplace: Literal[False] = ...) -> int: ...

@overload
def func(inplace: bool = ...) -> Optional[int]: ...


def func(inplace: bool = False): ...

inplace: bool
reveal_type(func(inplace=inplace))
```
it (correctly) gives us
```python
main.py:16: note: Revealed type is 'Union[builtins.int, None]'
```

Reply:
sure, will update.  for my edification, why isn't this caught by the regular mypy run?

Reply:
thanks - as far as I understand, if you have any overloads, then mypy will only look at them and not at the function signature

The reason you sometimes don't need to do this is that, according to [the docs](https://mypy.readthedocs.io/en/stable/more_types.html):

> if multiple variants match due to one or more of the arguments being a union, mypy will make the inferred type be the union of the matching variant returns:

So, if one overload is `str` and another is `bool` and you pass something that's `Union[str, bool]`, it'll infer the return type as being the union of the two overloads

Here, however, that is not the case, because it doesn't recognise `bool` as being the union of `Literal[True]` and `Literal[False]`